src/Microsoft.Health.Dicom.Core/Features/Workitem/WorkitemOrchestrator.cs

```diff
@@ -78,10 +81,86 @@ await StoreWorkitemBlobAsync(identifier, dataset, cancellationToken)
             }
         }
 
+        public async Task CancelWorkitemAsync(string workitemInstanceUid, DicomDataset dataset, CancellationToken cancellationToken)
+        {
+            EnsureArg.IsNotNull(dataset, nameof(dataset));
+            EnsureArg.IsNotNull(workitemInstanceUid, nameof(workitemInstanceUid));
+
+            try
+            {
+                var partitionKey = _contextAccessor.RequestContext.GetPartitionKey();
+
+                // TODO: Need to find a way to move this out of Orchestrator. Should be part of the main validator???
+                var futureProcedureStepState = await ValidateProcedureStepStateInStoreAsync(workitemInstanceUid, partitionKey, cancellationToken);
+
+                dataset.AddOrUpdate(DicomTag.ProcedureStepState, futureProcedureStepState);
+
+                var queryTags = await _workitemQueryTagService.GetQueryTagsAsync(cancellationToken).ConfigureAwait(false);
+                var workitemKey = await _indexWorkitemStore
+                    .UpdateWorkitemAsync(partitionKey, workitemInstanceUid, dataset, queryTags, cancellationToken)
+                    .ConfigureAwait(false);
+
+                var workitemInstanceIdentifier = new WorkitemInstanceIdentifier(workitemInstanceUid, workitemKey, partitionKey);
+
+                dataset.TryGetString(DicomTag.ReasonForCancellation, out var cancellationReason);
+                if (!string.IsNullOrWhiteSpace(cancellationReason))
+                {
+                    var blobDicomDataset = await GetWorkitemBlobAsync(workitemInstanceIdentifier, cancellationToken).ConfigureAwait(false);
+
```

Comment:
So if there is any issue with blob storage, we still want to proceed with the db update or do we need to revert?

Reply:
valid question. let's discuss on how we want to handle this.

Reply:
Implemented the changes based on our discussion.